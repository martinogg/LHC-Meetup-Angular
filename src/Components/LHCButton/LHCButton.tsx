
import React, { Component, Children } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';

interface IProps {
    onSelected(): void
    styles?: any
}

export default class LHCButton extends Component<IProps> {

    constructor(props: IProps) {

        super(props);
    }

    render() {

        return <TouchableOpacity style={[localStyles.container, this.props.styles]} onPress={() => this.props.onSelected()}>
            {this.props.children}
        </TouchableOpacity>
    }
}

const localStyles = StyleSheet.create({
    container: {
        margin: 10,
        backgroundColor: '#2980b6',
        paddingVertical: 15
    },
})