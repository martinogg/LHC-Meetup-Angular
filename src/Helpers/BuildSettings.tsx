const jsonkey = {
    LHC: {
        FIREBASEapiKey: "FIREBASEKEY",
        FIREBASEauthDomain: "FIREBASEKEY",
        FIREBASEdatabaseURL: "FIREBASEKEY",
        FIREBASEstorageBucket: "FIREBASEKEY",
    }
};

const build = 'LHC';

class BuildSettings {
    
    public static get(value: string): any {
        const app = jsonkey[build]
        return app[value]
    }

}

export default BuildSettings;