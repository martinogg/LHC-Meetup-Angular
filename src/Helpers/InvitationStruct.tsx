export interface IInvitation {
    from: string,
    to: string,
    reason: string
}

export interface IInvitationFromFirebase {
    id: string,
    invitation: IInvitation
}

export interface IInvitationFromAndTo {
    from: IInvitationFromFirebase[],
    to: IInvitationFromFirebase[]
}

export class Invitation {

    public static create(_from: string, _to: string, _reason: string): IInvitation {

        const ret = {
            from: _from,
            to: _to,
            reason: _reason
        }
        return ret
    }
}