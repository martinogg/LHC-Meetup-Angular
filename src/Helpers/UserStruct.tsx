export interface IUserInterest {
    title: string,
    description: string
}

export interface IUser {
    userName: string,
    userLocation: string,
    userContact: string,
    userInterests: IUserInterest[],
}

export interface IUserFromFirebase {
    id: string,
    user: IUser
}

export class User {

    public static create(name: string, location: string, contact: string, interests: IUserInterest[]): IUser {

        const ret: IUser = {
            userName: name,
            userLocation: location,
            userContact: contact,
            userInterests: interests
        }

        return ret
    }
}

export class UserInterest {

    public static create(_title: string, _description: string): IUserInterest {

        const ret: IUserInterest = {
            title: _title,
            description: _description
        }

        return ret
    }
}