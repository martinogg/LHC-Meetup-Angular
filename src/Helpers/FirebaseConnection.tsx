// TODO Refactor this mess and get some tests done

'use strict'
import React from 'react';
import {
  Alert,
} from 'react-native';
import * as firebase from 'firebase';
import 'firebase/firestore';
import BuildSettings from './BuildSettings';

import { IUser, User, IUserFromFirebase, IUserInterest } from './UserStruct'
import { IInvitation, IInvitationFromFirebase, Invitation, IInvitationFromAndTo, InvitationStatus, IInvitationFromAndToWithUserObjects, IInvitationFromFirebaseWithUserObject } from './InvitationStruct'

var instance: FirebaseConnection | null = null;

class FirebaseConnection {
  firebaseApp: firebase.app.App | null = null;

  static getInstance(): FirebaseConnection {
    if (instance) {
      return instance;
    } else {
      instance = new FirebaseConnection();
      instance.initApp();
      return instance;
    }
  }

  initApp() {
    var firebaseConfig = {
      apiKey: BuildSettings.get("FIREBASEapiKey"),
      authDomain: BuildSettings.get("FIREBASEauthDomain"),
      databaseURL: BuildSettings.get("FIREBASEdatabaseURL"),
      storageBucket: BuildSettings.get("FIREBASEstorageBucket"),
      projectId: BuildSettings.get("FIREBASEprojectId")
    };

    if (firebaseConfig.apiKey != "") {
      this.firebaseApp = firebase.initializeApp(firebaseConfig);
    }
  }

  getProperties(obj, count) {
    if (isNaN(count)) count = Infinity
    var keys = []
    for (var it in obj) {
      if (keys.length > count) break;
      keys.push(it);
    }
    return keys;
  }

  public logout(): Promise<void> {

    return new Promise((resolve, reject) => {
      firebase.auth().signOut().then(() => {

        resolve()
      }, (failReason) => {

        Alert.alert('Error: ' + failReason)
        resolve()
      })
    })
  }

  // TODO TEST
  public checkFBVersion(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {

      if (firebase.auth().currentUser == null) {

        reject('User Not Logged in')
        return
      }

      const localVersion: number = BuildSettings.get("FIREBASEVersion")

      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-settings").get()
        .then((snap) => {

          let fbVersion: number = 99999999
          snap.forEach(setting => {

            if (setting.id == 'version') {

              const data = setting.data()
              fbVersion = +data.version
            }

          });

          resolve(localVersion >= fbVersion)
        })
        .catch(function (error) {

          reject(error)
        });
    })
}


  //TODO TEST
  public buildInvitationsWithUserObjects(invitations: IInvitationFromAndTo): Promise < IInvitationFromAndToWithUserObjects > {

  return new Promise<IInvitationFromAndToWithUserObjects>((resolve, reject) => {

    this.allUsers().then((users) => {

      let fromInvitations: IInvitationFromFirebaseWithUserObject[] = []
      let toInvitations: IInvitationFromFirebaseWithUserObject[] = []

      const createNewInvitation: (invitation: IInvitationFromFirebase) => IInvitationFromFirebaseWithUserObject = (invitation) => {

        const newInvitation: IInvitationFromFirebaseWithUserObject = {

          ...invitation,
          fromObject: this.userObjectForID(invitation.invitation.from, users),
          toObject: this.userObjectForID(invitation.invitation.to, users)
        }

        return newInvitation
      }

      invitations.from.forEach((element) => {

        fromInvitations.push(createNewInvitation(element))
      })

      invitations.to.forEach((element) => {

        toInvitations.push(createNewInvitation(element))
      })

      resolve({ from: fromInvitations, to: toInvitations })

    }).catch((error) => {

      reject(error)
    })

  })
}

  private userObjectForID(id: string, users: IUserFromFirebase[]): IUserFromFirebase {

  let ret: IUserFromFirebase | null = null
  users.forEach((element) => {

    if (element.id == id) {

      ret = element
    }
  })

  if (ret == null) {

    ret = { id: id, user: User.create('Unknown User', '', '', []) }
  }

  return ret
}

  public getInvitations(): Promise < IInvitationFromAndTo > {
  // TODO TEST
  return new Promise<IInvitationFromAndTo>((resolve, reject) => {

    if (firebase.auth().currentUser == null) {

      reject('User Not Logged in')
      return
    }

    const currentUser = firebase.auth().currentUser as firebase.User

    firebase.firestore().collection("LHC-Invitations").get()
      .then((snap) => {

        let ret: IInvitationFromFirebase[] = []

        snap.forEach(invitationInList => {

          const id = invitationInList.id
          const data = invitationInList.data()

          if (data) {
            
            const invitation = Invitation.create(data.from, data.to, data.reason, data.status)
            ret.push({ id: id, invitation: invitation })
          }

        });

        resolve({ from: ret.filter((item) => { return item.invitation.from == currentUser.uid }), to: ret.filter((item) => { return item.invitation.to == currentUser.uid }) })
      })
      .catch(function (error) {

        reject(error)
      });
  })

}

  public deleteInvitation(uid: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).delete()
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public updateInvitationResponse(uid: string, response: InvitationStatus): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ status: response })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public updateInvitation(uid: string, reason: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ reason })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public createNewInvitation(invitation: IInvitation): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      const newInvitationId = this.dateEpochPlusUID()
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Invitations").doc(newInvitationId).set(invitation)
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public saveUserDetails(user: IUser): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).set({
        name: user.userName,
        location: user.userLocation,
        contact: user.userContact,
        interests: user.userInterests,
      })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  private queryMatchesInterest(query: string, interestList: IUserInterest[]): boolean {

  let found = false
  const queryLower = query.toLowerCase()
  interestList.forEach((interest) => {

    if (interest.title.toLowerCase().includes(queryLower)) {

      found = true
    }
    if (interest.description.toLowerCase().includes(queryLower)) {

      found = true
    }
  })

  return found
}

  public searchOtherUsers(query: string): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()
            const matchesInterest = this.queryMatchesInterest(query, data.interests)

            if (data != null && id != currentUser.uid && matchesInterest) {

              const user = User.create(data.name, data.location, data.contact, data.interests)
              ret.push({ id: id, user: user })
            }

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public allUsers(): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()

            const user = User.create(data.name, data.location, data.contact, data.interests)
            ret.push({ id: id, user: user })

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public loadUserDetails(): Promise < IUser > {
  // TODO TEST
  return new Promise<IUser>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).get()
        .then((snap) => {

          const data = snap.data()

          if (data != null) {
            const user = User.create(data.name, data.location, data.contact, data.interests)

            resolve(user)
          } else {
            reject('An Error Occurred')
          }
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public login(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().signInWithEmailAndPassword(username, password).then(() => {

      resolve()
    }, (reason) => {

      reject(reason)
    })
  })
}

  public register(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().createUserWithEmailAndPassword(username, password).then(() => {


      // TODO - Shift this out in to separate email verification
      /*
      if (this.isLoggedIn()) {
 
        firebase.auth().currentUser.sendEmailVerification()
      }
      */

      resolve()
    }, (reason) => {
      reject(reason)
    })
  })
}

  public isLoggedIn(): boolean {
  // TODO TEST

  return firebase.auth().currentUser != null
}

  public getCurrentUserID(): string {
  // TODO TEST
  if (firebase.auth().currentUser == null) {
    return 'Error: Not logged In'
  }

  const currentUser = firebase.auth().currentUser as firebase.User
  return currentUser.uid
}

  private dateEpochPlusUID(): string {
  // TODO TEST
  return (new Date).getTime().toString() + '-' + (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1)
}

sendComment(comment: string) {
  // TODO TEST

  if (comment !== '') {

    const date = this.dateEpochPlusUID()
    if (firebase.auth().currentUser != null) {

      const currentUser = firebase.auth().currentUser as firebase.User
      firebase.firestore().collection("LHC-Comments").doc(date).set({
        'comment': comment
      })
    }
  }
}

loginAnon() {
  let strongThis = this;
  if (this.firebaseApp != null) {
    firebase.auth().signInAnonymously().catch((error) => { });
  }
}

// only allows callback if user is actually logged in
_getCurrentUser(onGetCurrentUserFunc) {

  if (this.firebaseApp != null) {

    var currentUser = firebase.auth().currentUser;
    if (currentUser) {

      onGetCurrentUserFunc(currentUser);
    } else {

      // try again just in case user is back online
      this.loginAnon();
    }
  }
}

getLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);
    ref.once('value', (snap) => {
      var isSet = snap.val() === true;
      callback(isSet);
    });
  });
}

toggleLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.once('value', (snap) => {
      var alreadySet = snap.val() === true;
      var newSet = alreadySet ? null : true;
      ref.set(newSet).then(() => {
        callback(newSet);
      });
    });
  });
}

setViewedWithId(idStr, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("views").child(idStr);

    ref.set(true).then(() => {
      onCompleteFunc();
    });
  });
}

setLikeWithId(idStr, liked, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.set(liked ? true : null).then(() => {
      onCompleteFunc();
    });
  });
}

setDownloadedWithId(idStr) {
  this._getCurrentUser((currentUser) => {
    this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("downloads").child(idStr).set(true);
  });
}

getDbDecks(onCompleteFunc) {

  if (this.firebaseApp != null) {

    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = [];
      snap.forEach((child) => {
        items.push(child);
      });
      onCompleteFunc(items);
    });
  }
}

getDbDictLikes(onCompleteFunc) {

  if (this.firebaseApp != null) {
    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = {};
      snap.forEach((child) => {
        var likesCount = 0 + child.val().likes;
        var viewsCount = 0 + child.val().views;
        items["" + child.key] = { likes: likesCount, views: viewsCount };
      });
      onCompleteFunc(items);
    });
  }
}



alert(message) {
  Alert.alert(
    'Error',
    'Error:' + message,
    [
      { text: 'OK', onPress: () => console.log('OK Pressed') },
    ],
    { cancelable: false }
  );
}

}

export default FirebaseConnection // TODO FirebaseConnection is old style, remove this in favor of REDUX Version in the future

export const firebaseService = FirebaseConnection.getInstance().firebaseApp as firebase.app.App