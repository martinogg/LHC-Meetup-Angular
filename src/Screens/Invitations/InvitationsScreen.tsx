import React, { Component } from 'react'
import { Button, FlatList, Platform, StyleSheet, Text, View, Alert } from 'react-native'
import { NavigationActions, NavigationScreenProp, StackActions } from 'react-navigation'
import { AppStyles } from '../../AppStyles'
import FirebaseConnection from '../../Helpers/FirebaseConnection'
import { IInvitationFromFirebase } from '../../Helpers/InvitationStruct';

interface IProps {

    navigation: NavigationScreenProp<any, any>,
    screenProps: {
        firebaseConnection: FirebaseConnection
    }
}

interface IState {
    sentInvitations: IInvitationFromFirebase[]
    receivedInvitations: IInvitationFromFirebase[]
}

export class InvitationsScreen extends Component<IProps, IState> {

    private static navigationOptions = {
        title: 'Invitations'
    }

    constructor(props: IProps) {
        super(props)

        this.state = {

            sentInvitations: [],
            receivedInvitations: []
        }
    }

    public componentDidMount() {

        this.getInvitations()
    }

    private getInvitations() {

        this.props.screenProps.firebaseConnection.getInvitations().then((_invitations) => {

            this.setState({ sentInvitations: _invitations.from, receivedInvitations: _invitations.to })
        }, (error) => {

            Alert.alert('Error:' + error)
        })
    }

    private invitationComponentForElement(item: IInvitationFromFirebase): JSX.Element {

        // TODO - this needs to look better
        return <View>
            <Text>ID{item.id}</Text>
            <Text>FROM{item.invitation.from}</Text>
            <Text>TO{item.invitation.to}</Text>
            <Text>REASON{item.invitation.reason}</Text>
            <Text>.</Text>
        </View>
    }

    private listElements(sentInvitations: IInvitationFromFirebase[], receivedInvitations: IInvitationFromFirebase[]): JSX.Element[] {

        let ret: JSX.Element[] = []

        if (sentInvitations.length > 0) {
            ret.push(<Text style={AppStyles.buttonText}>Sent</Text>)
        }

        sentInvitations.forEach(element => {
            ret.push(this.invitationComponentForElement(element))
        });


        if (receivedInvitations.length > 0) {
            ret.push(<Text style={AppStyles.buttonText}>Received</Text>)
        }

        receivedInvitations.forEach(element => {
            ret.push(this.invitationComponentForElement(element))
        });

        if (ret.length == 0) {

            ret.push(<Text>No Invitations. Go to Browse Screen to find a new person to meet</Text>)
        }
        return ret
    }

    public render() {

        const listElements = this.listElements(this.state.sentInvitations, this.state.receivedInvitations)

        return (
            <View style={AppStyles.container}>
                <FlatList
                    data={listElements}
                    keyExtractor={(item, index) => index.toString()}
                    renderItem={({ item }) => { return item }}
                />
            </View>
        )
    }
}