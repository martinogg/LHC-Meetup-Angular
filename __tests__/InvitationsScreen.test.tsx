import { shallow } from 'enzyme';
import React from 'react';
import { Alert } from 'react-native';
import LHCButton from '../src/Components/LHCButton/LHCButton'

import { IInvitationFromAndTo, IInvitation, Invitation, IInvitationFromFirebase } from '../src/Helpers/InvitationStruct'
import { InvitationsScreen } from '../src/Screens/Invitations/InvitationsScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { tsModuleBlock } from '@babel/types';

const createTestProps = (props: Object) => ({
    navigation: {
        navigate: jest.fn(),
        replace: jest.fn(),
        dispatch: jest.fn()
    },
    screenProps: {
        firebaseConnection: {
            isLoggedIn: () => { return true },
            getInvitations: () => { return new Promise<IInvitationFromAndTo>((resolve, reject) => { resolve({ to: [], from: [] }) }) }
        }
    },
    ...props
});

it('should display HomeScreen with no errors', () => {

    const props = createTestProps({})
    const navigation = { navigate: jest.fn() };

    expect(renderer.create(<InvitationsScreen {...props} />)).toMatchSnapshot();
});

test('componentDidMount function', () => {

    const props = createTestProps({})
    const sut = new InvitationsScreen(props)

    sut.getInvitations = jest.fn()

    sut.componentDidMount()

    expect(sut.getInvitations).toHaveBeenCalledTimes(1)
})

test('getInvitations function Success', async () => {

    const testFromList: IInvitationFromFirebase[] = [{ id: '1', invitation: Invitation.create('a', 'b', 'c') }, { id: '2', invitation: Invitation.create('d', 'e', 'f') }]
    const testToList: IInvitationFromFirebase[] = [{ id: '12', invitation: Invitation.create('a2', 'b2', 'c2') }, { id: '22', invitation: Invitation.create('d2', 'e2', 'f2') }]

    const props = createTestProps({

        screenProps: {
            firebaseConnection: {
                getInvitations: () => { return new Promise<IInvitationFromAndTo>((resolve, reject) => { resolve({ to: testToList, from: testFromList }) }) }
            }
        },

    })
    const sut = new InvitationsScreen(props)

    sut.setState = jest.fn()

    await sut.getInvitations()

    expect(sut.setState).toHaveBeenCalledTimes(1)
    expect(sut.setState).toHaveBeenCalledWith({ "receivedInvitations": [{ "id": "12", "invitation": { "from": "a2", "reason": "c2", "to": "b2" } }, { "id": "22", "invitation": { "from": "d2", "reason": "f2", "to": "e2" } }], "sentInvitations": [{ "id": "1", "invitation": { "from": "a", "reason": "c", "to": "b" } }, { "id": "2", "invitation": { "from": "d", "reason": "f", "to": "e" } }] })
})

test('getInvitations function Fail', async () => {

    const props = createTestProps({

        screenProps: {
            firebaseConnection: {
                getInvitations: () => {
                    return new Promise<IInvitationFromAndTo>((resolve, reject) => { reject('anError') })
                }
            }
        }

    })
    const sut = new InvitationsScreen(props)

    jest.resetAllMocks()

    jest.mock('Alert', () => {
        return {
            alert: jest.fn()
        }
    });

    await sut.getInvitations()

    expect(Alert.alert).toHaveBeenCalledTimes(1)
    expect(Alert.alert).toHaveBeenCalledWith('Error:anError')

})

it('should display HomeScreen with no Entries', async () => {

    const props = createTestProps({
        screenProps: {
            firebaseConnection: {
                getInvitations: () => { return new Promise<IInvitationFromAndTo>((resolve, reject) => { resolve({ to: [], from: [] }) }) }
            }
        },
    })

    const wrapper = shallow(<InvitationsScreen {...props} />)
    const sut: any = wrapper.instance()

    await sut.componentDidMount()
    const renderOutput = sut.render()
    expect(renderOutput).toMatchSnapshot();

});

it('should display HomeScreen with Entries', async () => {

    const testFromList: IInvitationFromFirebase[] = [{ id: '1', invitation: Invitation.create('a', 'b', 'c') }, { id: '2', invitation: Invitation.create('d', 'e', 'f') }]
    const testToList: IInvitationFromFirebase[] = [{ id: '12', invitation: Invitation.create('a2', 'b2', 'c2') }, { id: '22', invitation: Invitation.create('d2', 'e2', 'f2') }]

    const props = createTestProps({

        screenProps: {
            firebaseConnection: {
                getInvitations: () => { return new Promise<IInvitationFromAndTo>((resolve, reject) => { resolve({ to: testToList, from: testFromList }) }) }
            }
        },

    })

    const wrapper = shallow(<InvitationsScreen {...props} />)
    const sut: any = wrapper.instance()
    await sut.componentDidMount()
    const renderOutput = sut.render()
    expect(renderOutput).toMatchSnapshot();
});