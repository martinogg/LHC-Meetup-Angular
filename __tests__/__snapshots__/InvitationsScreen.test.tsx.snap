// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display HomeScreen with Entries 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <FlatList
    data={
      Array [
        <Text
          style={
            Object {
              "color": "#fff",
              "fontWeight": "700",
              "textAlign": "center",
            }
          }
        >
          Sent
        </Text>,
        <View>
          <Text>
            ID
            1
          </Text>
          <Text>
            FROM
            a
          </Text>
          <Text>
            TO
            b
          </Text>
          <Text>
            REASON
            c
          </Text>
          <Text>
            .
          </Text>
        </View>,
        <View>
          <Text>
            ID
            2
          </Text>
          <Text>
            FROM
            d
          </Text>
          <Text>
            TO
            e
          </Text>
          <Text>
            REASON
            f
          </Text>
          <Text>
            .
          </Text>
        </View>,
        <Text
          style={
            Object {
              "color": "#fff",
              "fontWeight": "700",
              "textAlign": "center",
            }
          }
        >
          Received
        </Text>,
        <View>
          <Text>
            ID
            12
          </Text>
          <Text>
            FROM
            a2
          </Text>
          <Text>
            TO
            b2
          </Text>
          <Text>
            REASON
            c2
          </Text>
          <Text>
            .
          </Text>
        </View>,
        <View>
          <Text>
            ID
            22
          </Text>
          <Text>
            FROM
            d2
          </Text>
          <Text>
            TO
            e2
          </Text>
          <Text>
            REASON
            f2
          </Text>
          <Text>
            .
          </Text>
        </View>,
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`should display HomeScreen with no Entries 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <FlatList
    data={
      Array [
        <Text>
          No Invitations. Go to Browse Screen to find a new person to meet
        </Text>,
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`should display HomeScreen with no errors 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        <Text>
          No Invitations. Go to Browse Screen to find a new person to meet
        </Text>,
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          No Invitations. Go to Browse Screen to find a new person to meet
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
